vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/step/DescendantOrSelfPathStepUnnesting.java

```diff
@@ -100,13 +121,12 @@ private boolean processNodeTree(TaggedValuePointable rootTVP, IPointable result)
             tvpItem.set(rootTVP);
             try {
                 setNodeToResult(tvpItem, result);
-                return true;
```

Comment:
The logic is wrong here. You only want to pass one node result foreach step call. This will  pass more than one and does not consider the filter.